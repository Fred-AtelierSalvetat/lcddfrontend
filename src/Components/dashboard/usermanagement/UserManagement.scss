$line-separator-color: #eceeef;
$headers-color: #eceeef;
$header-background-color: #f2f2f2;
$input-border-color: #cccccc;
$input-place-holder-color: #999999;
$awaiting-validation-background-color: #f0ad4e;

#lcdd-dashboard-page {
    .navPill {
        &-enter {
            background-color: yellowgreen !important;
        }

        &-enter-active {
            background-color: blue !important;
            transition: background-color 2000ms !important;
        }
        &-exit {
            opacity: 1;
            background-color: yellow !important;
            transition: opacity 1000ms;
        }

        &-exit-active {
            opacity: 1;
            background-color: yellow !important;
            transition: opacity 1000ms;
        }
    }
    .table_and_filters_pane {
        background-color: $header-background-color;
        border-top-left-radius: 0.25rem;
        border-top-right-radius: 0.25rem;
        //width: fit-content;
        min-width: 100%;
        filter: drop-shadow(0 0.25rem 0.5rem rgba(130, 130, 130, 0.5));
        .nav {
            padding: 0.5rem;

            .nav-link {
                transition: transform 0.2s;
                padding: 0.5rem;
                color: var(--primary);
                background-color: transparent;
                &:hover {
                    transform: scale(1.1);
                }
                &.nav-link.active {
                    color: var(--white);
                    background-color: var(--primary);
                    &:hover {
                        transform: none;
                    }
                }
            }
            .searchbox {
                width: calc(100vw - 160px);
                position: relative;
                align-self: bottom;
                //TODO
                @media (min-width: 769px) {
                    margin-left: auto;
                    min-width: 200px;
                    width: 400px;
                    align-self: top;
                }
                input {
                    padding: 0.5rem 1rem;
                    border: 0;
                    font-size: 1rem;
                    font-style: normal;
                    font-weight: 400;
                    line-height: 1.5rem;
                    letter-spacing: 0;
                    text-align: left;
                    border: 1px solid $input-border-color;
                    &::placeholder {
                        color: $input-place-holder-color;
                    }
                }

                svg {
                    position: absolute;
                    right: 10px;
                    top: 13px;
                    z-index: 9999;
                }
            }
        }

        .tab-content {
            background-color: var(--white);
            margin-bottom: 7px;
            .tab-pane {
                height: calc(100vh - 200px);
                min-height: 97px;
                overflow: auto;
            }
            thead {
                tr {
                    th {
                        position: sticky;
                        top: 0;
                        z-index: 10;
                        background-color: var(--gray-dark);
                        color: $headers-color;
                        padding: 0.75rem 1rem;
                        font-style: normal;
                        font-weight: 700;
                        line-height: 1, 5rem;
                        letter-spacing: 0;
                        text-align: left;
                        vertical-align: middle;
                        white-space: nowrap;
                        > .CenteredHeader {
                            text-align: center;
                        }
                    }
                }
            }
            tbody {
                tr {
                    border-top: 1px solid $line-separator-color;
                    td {
                        z-index: 9;
                        padding: 0.75rem 1rem;
                        background-color: var(--white);
                        .Centered {
                            display: flex;
                            align-items: center;
                            justify-content: center;

                            .speakerStatus {
                                color: var(--white);
                                border-radius: 20px;

                                height: 20px;
                                width: 68px;
                                display: flex;
                                align-items: center;
                                justify-content: center;
                                margin: 4px 0 6px 0;
                                &.speakerAwaitingAnswer {
                                    background-color: var(--cyan);
                                }
                                &.speakerAwaitingValidation {
                                    background-color: $awaiting-validation-background-color;
                                }
                                p {
                                    margin: 0;
                                    line-height: 22px;
                                    font-size: 12px;
                                    font-style: normal;
                                    font-weight: 700;
                                    letter-spacing: 0px;
                                    white-space: nowrap;
                                }
                            }
                        }

                        > p {
                            margin: 0;
                            font-style: normal;
                            font-weight: 400;
                            line-height: 1.5rem;
                            letter-spacing: 0px;
                            text-align: left;
                            vertical-align: middle;
                            white-space: nowrap;
                        }
                        > a {
                            font-size: small;
                            font-weight: 400;
                            font-style: normal;
                            white-space: nowrap;
                            color: var(--link);
                        }
                        .ActionMenuIcon {
                            display: flex;
                            align-items: center;
                            justify-content: center;
                            cursor: pointer;
                        }
                        .userStatus {
                            display: flex;
                            align-items: center;
                            justify-content: center;
                            > svg {
                                cursor: pointer;

                                &:hover {
                                    transform: scale(1.1);
                                }
                                #toggle_on_24px {
                                    fill: var(--success);
                                }
                                #toggle_off_24px {
                                    fill: var(--danger);
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}

.arrow {
    display: none !important;
}

.popover-body {
    margin: 0;
    padding: 0 !important;
    border-width: 1px;
    display: flex;
    flex-direction: column;
    > .actionButton {
        display: flex;
        align-items: center;
        padding: 0.5rem 1rem;
        border-bottom: 1px solid var(--gray-light);
        cursor: pointer;

        &:nth-last-child(1) {
            border-bottom: none;
        }

        .actionButtonPanel {
            transition: transform 0.2s;
            &:hover {
                transform: scale(1.1);
            }

            span {
                height: 1.5rem;
                margin: 0;
                padding-left: 0.5rem;
                font-weight: 400;
                font-style: normal;
            }
        }
    }
}
