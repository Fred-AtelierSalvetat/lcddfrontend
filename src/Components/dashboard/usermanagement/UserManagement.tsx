import React, { FC, useEffect } from 'react';
import { useParams, Redirect, useLocation } from 'react-router-dom';
import { useSelector, useDispatch } from 'react-redux';

import Nav from 'react-bootstrap/Nav';
import Tab from 'react-bootstrap/Tab';
import Table from 'react-bootstrap/Table';
import Button from 'react-bootstrap/Button';

import { ReactComponent as ToggleOnIcon } from '../../../assets/icons/toggle_on_24px.svg';
import { ReactComponent as ToggleOffIcon } from '../../../assets/icons/toggle_off_24px.svg';
import { ReactComponent as DeleteForeverIcon } from '../../../assets/icons/delete_forever_16px.svg';
import { ReactComponent as AdminIcon } from '../../../assets/icons/admin_24px.svg';
import { ReactComponent as ValidateIcon } from '../../../assets/icons/validate_24px.svg';
import { ReactComponent as InviteSpeakerIcon } from '../../../assets/icons/group_add_24px.svg';

import * as userRoles from '../../../state/users/constants/Roles';
import * as userStatus from '../../../state/users/constants/Status';
import * as actionTypes from '../../../state/users/constants/ActionTypes';
import * as usersAction from '../../../state/users/actions';
import { getVisibleUsers, isRequestInProgress } from '../../../state/reducers';
import ErrorBoundary from '../ErrorBoundary';

import SearchBox from './Searchbox';
import ActionMenuCell from './ActionMenuCell';
import Action from './Action';
import ConfirmDialog from '../../shared/modals/ConfirmDialog';

import './UserManagement.scss';

const ADMIN_TAB_ROLE_FILTER = [userRoles.ROLE_ADMIN];
const SPEAKER_TAB_ROLE_FILTER = [
    userRoles.ROLE_SPEAKER_AWAITING_ANSWER,
    userRoles.ROLE_SPEAKER_AWAITING_VALIDATION,
    userRoles.ROLE_SPEAKER,
];
const USERS_TAB_ROLE_FILTER = [
    userRoles.ROLE_PRO_USER,
    userRoles.ROLE_STUDENT,
    userRoles.ROLE_CITIZEN,
    userRoles.ROLE_SPEAKER_AWAITING_ANSWER,
    userRoles.ROLE_SPEAKER_AWAITING_VALIDATION,
    userRoles.ROLE_SPEAKER,
];

const roleFilterMap = {
    admin: ADMIN_TAB_ROLE_FILTER,
    speaker: SPEAKER_TAB_ROLE_FILTER,
    user: USERS_TAB_ROLE_FILTER,
};

const UserManagement: FC = () => {
    const { roleFilter } = useParams() as {
        roleFilter?: string;
    };

    const dispatch = useDispatch();

    useEffect(() => {
        const fetchAction = usersAction.fetchUsers({
            failureAlertMsg: (
                <div className="JSXalertMsg">
                    <span>Echec du des données utilisateurs</span>
                    <Button
                        className="retry"
                        onClick={() => {
                            dispatch(fetchAction);
                        }}
                    >
                        Relancer
                    </Button>
                </div>
            ),
        });
        dispatch(fetchAction);
    }, [dispatch]); //TODO Define refresh strategy

    let root_path = useLocation().pathname;
    roleFilter && (root_path = root_path.replace(new RegExp('/' + roleFilter + '$'), ''));

    const isFetching = useSelector(isRequestInProgress(actionTypes.FETCH_USERS_REQUEST));

    const users = useSelector(getVisibleUsers);

    if (!roleFilter) {
        return <Redirect to={root_path + '/admin'} />;
    }

    if (!Object.keys(roleFilterMap).includes(roleFilter)) {
        return <Redirect to="/no-match" />;
    }

    dispatch(usersAction.setUsersRoleFilter(roleFilterMap[roleFilter]));

    const tabs_desc: {
        tab_label: string;
        uri_filter: string;
        table_columns: {
            key: string;
            renderHeader: () => JSX.Element;
            renderCell: any;
        }[];
    }[] = [
        {
            tab_label: 'Admins',
            uri_filter: 'admin',
            table_columns: [
                {
                    key: 'lastname',
                    renderHeader: () => <div>Nom</div>,
                    renderCell: ({ lastname }) => <p>{lastname}</p>,
                },
                {
                    key: 'firstname',
                    renderHeader: () => <div>Prénom</div>,
                    renderCell: ({ firstname }) => <p>{firstname}</p>,
                },
                { key: 'phone', renderHeader: () => <div>Téléphone</div>, renderCell: ({ phone }) => <p>{phone}</p> },
                {
                    key: 'email_pro',
                    renderHeader: () => <div>E-mail Pro</div>,
                    renderCell: ({ email_pro }) => <p>{email_pro}</p>,
                },
                {
                    key: 'email',
                    renderHeader: () => <div>E-mail Perso</div>,
                    renderCell: ({ email }) => <p>{email}</p>,
                },
                { key: 'town', renderHeader: () => <div>Ville</div>, renderCell: ({ town }) => <p>{town}</p> },
                {
                    key: 'status',
                    renderHeader: () => <div className="CenteredHeader">Status</div>,

                    renderCell: ({ user_id, status }) => (
                        <div className="userStatus">
                            {status === userStatus.ACTIVE && (
                                <ToggleOnIcon onClick={() => dispatch(usersAction.deactivateUser(user_id))} />
                            )}
                            {status === userStatus.INACTIVE && (
                                <ToggleOffIcon onClick={() => dispatch(usersAction.activateUser(user_id))} />
                            )}
                        </div>
                    ),
                },
                {
                    key: 'actions',
                    renderHeader: () => <div className="CenteredHeader">Actions</div>,
                    renderCell: ({ user_id }) => {
                        return (
                            <ActionMenuCell>
                                <Action
                                    icon={<AdminIcon />}
                                    label="Revenir intervenant"
                                    action={() => dispatch(usersAction.revokeUserAdminRight(user_id))}
                                />
                                <Action
                                    icon={<DeleteForeverIcon />}
                                    label="Supprimer"
                                    modalConfirmation={
                                        <ConfirmDialog
                                            show={true}
                                            title="Supprimer cet utilisateur"
                                            body="Cette action n’est pas réversible."
                                            cancelButton="Annuler"
                                            okButton="Supprimer"
                                            handleClose={() => {}}
                                            handleConfirm={() => dispatch(usersAction.deleteUser(user_id))}
                                        />
                                    }
                                />
                            </ActionMenuCell>
                        );
                    },
                },
                {
                    key: 'profile',
                    renderHeader: () => <div></div>,
                    renderCell: ({ user_id }) => <a href={'#/profile/' + user_id}>Voir profil</a>,
                },
            ],
        },
        {
            tab_label: 'Intervenants',
            uri_filter: 'speaker',
            table_columns: [
                {
                    key: 'lastname',
                    renderHeader: () => <div>Nom</div>,
                    renderCell: ({ lastname }) => <p>{lastname}</p>,
                },
                {
                    key: 'firstname',
                    renderHeader: () => <div>Prénom</div>,
                    renderCell: ({ firstname }) => <p>{firstname}</p>,
                },
                { key: 'phone', renderHeader: () => <div>Téléphone</div>, renderCell: ({ phone }) => <p>{phone}</p> },
                {
                    key: 'email_pro',
                    renderHeader: () => <div>E-mail Pro</div>,
                    renderCell: ({ email_pro }) => <p>{email_pro}</p>,
                },
                {
                    key: 'email',
                    renderHeader: () => <div>E-mail Perso</div>,
                    renderCell: ({ email }) => <p>{email}</p>,
                },
                { key: 'town', renderHeader: () => <div>Ville</div>, renderCell: ({ town }) => <p>{town}</p> },
                {
                    key: 'status',
                    renderHeader: () => <div className="CenteredHeader">Status</div>,
                    renderCell: ({ user_id, role, status }) => {
                        return (
                            <div className="Centered">
                                {role === userRoles.ROLE_SPEAKER_AWAITING_ANSWER && (
                                    <div className="speakerStatus speakerAwaitingAnswer">
                                        <p>En attente</p>
                                    </div>
                                )}
                                {role === userRoles.ROLE_SPEAKER_AWAITING_VALIDATION && (
                                    <div className="speakerStatus speakerAwaitingValidation">
                                        <p>A valider</p>
                                    </div>
                                )}
                                {role !== userRoles.ROLE_SPEAKER_AWAITING_ANSWER &&
                                    role !== userRoles.ROLE_SPEAKER_AWAITING_VALIDATION && (
                                        <div className="userStatus">
                                            {status === userStatus.ACTIVE && (
                                                <ToggleOnIcon
                                                    onClick={() => dispatch(usersAction.deactivateUser(user_id))}
                                                />
                                            )}
                                            {status === userStatus.INACTIVE && (
                                                <ToggleOffIcon
                                                    onClick={() => dispatch(usersAction.activateUser(user_id))}
                                                />
                                            )}
                                        </div>
                                    )}
                            </div>
                        );
                    },
                },
                {
                    key: 'actions',
                    renderHeader: () => <div className="CenteredHeader">Actions</div>,
                    renderCell: ({ role, user_id }) => (
                        <ActionMenuCell>
                            {role === userRoles.ROLE_SPEAKER && (
                                <Action
                                    icon={<AdminIcon />}
                                    label="Promouvoir admin"
                                    action={() => dispatch(usersAction.promoteUserToAdmin(user_id))}
                                />
                            )}
                            {role === userRoles.ROLE_SPEAKER_AWAITING_VALIDATION && (
                                <Action
                                    icon={<ValidateIcon />}
                                    label="Valider candidat"
                                    action={() => dispatch(usersAction.validateSpeaker(user_id))}
                                />
                            )}
                            <Action
                                icon={<DeleteForeverIcon />}
                                label="Supprimer"
                                modalConfirmation={
                                    <ConfirmDialog
                                        show={true}
                                        title="Supprimer cet utilisateur"
                                        body="Cette action n’est pas réversible."
                                        cancelButton="Annuler"
                                        okButton="Supprimer"
                                        handleClose={() => {}}
                                        handleConfirm={() => dispatch(usersAction.deleteUser(user_id))}
                                    />
                                }
                            />
                        </ActionMenuCell>
                    ),
                },
                {
                    key: 'profile',
                    renderHeader: () => <div></div>,
                    renderCell: ({ user_id }) => <a href={'#/profile/' + user_id}>Voir profil</a>,
                },
            ],
        },
        {
            tab_label: 'Utilisateurs',
            uri_filter: 'user',
            table_columns: [
                {
                    key: 'lastname',
                    renderHeader: () => <div>Nom</div>,
                    renderCell: ({ lastname }) => <p>{lastname}</p>,
                },
                {
                    key: 'firstname',
                    renderHeader: () => <div>Prénom</div>,
                    renderCell: ({ firstname }) => <p>{firstname}</p>,
                },
                { key: 'email', renderHeader: () => <div>E-mail</div>, renderCell: ({ email }) => <p>{email}</p> },
                { key: 'town', renderHeader: () => <div>Ville</div>, renderCell: ({ town }) => <p>{town}</p> },
                {
                    key: 'role',
                    renderHeader: () => <div>Rôle</div>,
                    renderCell: ({ role }) => (
                        <p>
                            {role
                                .replace(userRoles.ROLE_SPEAKER_AWAITING_ANSWER, userRoles.ROLE_PRO_USER)
                                .replace(userRoles.ROLE_SPEAKER_AWAITING_VALIDATION, userRoles.ROLE_PRO_USER)
                                .replace(userRoles.ROLE_SPEAKER, userRoles.ROLE_PRO_USER)}
                        </p>
                    ),
                },

                {
                    key: 'status',
                    renderHeader: () => <div className="CenteredHeader">Status</div>,
                    renderCell: ({ user_id, status }) => (
                        <div className="userStatus">
                            {status === userStatus.ACTIVE && (
                                <ToggleOnIcon onClick={() => dispatch(usersAction.deactivateUser(user_id))} />
                            )}
                            {status === userStatus.INACTIVE && (
                                <ToggleOffIcon onClick={() => dispatch(usersAction.activateUser(user_id))} />
                            )}
                        </div>
                    ),
                },
                {
                    key: 'actions',
                    renderHeader: () => <div className="CenteredHeader">Actions</div>,
                    renderCell: ({ role, user_id }) => (
                        <ActionMenuCell>
                            {role === userRoles.ROLE_PRO_USER && (
                                <Action
                                    icon={<InviteSpeakerIcon />}
                                    label="Inviter à intervenir"
                                    action={() => dispatch(usersAction.promoteUserToSpeaker(user_id))}
                                />
                            )}
                            <Action
                                icon={<DeleteForeverIcon />}
                                label="Supprimer"
                                modalConfirmation={
                                    <ConfirmDialog
                                        show={true}
                                        title="Supprimer cet utilisateur"
                                        body="Cette action n’est pas réversible."
                                        cancelButton="Annuler"
                                        okButton="Supprimer"
                                        handleClose={() => {}}
                                        handleConfirm={() => dispatch(usersAction.deleteUser(user_id))}
                                    />
                                }
                            />
                        </ActionMenuCell>
                    ),
                },
            ],
        },
    ];

    return (
        <ErrorBoundary>
            {!users.length && isFetching ? (
                <p>Chargement...</p>
            ) : (
                <div className="table_and_filters_pane">
                    <Tab.Container>
                        <Nav variant="pills">
                            {tabs_desc.map((tab_desc) => {
                                return (
                                    <Nav.Item key={'tab_' + tab_desc.tab_label}>
                                        {/* <CSSTransition
                                            in={tab_desc.uri_filter === roleFilter}
                                            timeout={{ enter: 3000, exit: 3000 }}
                                            classNames="navPill"
                                        > */}
                                        <Nav.Link
                                            href={`#${root_path}/${tab_desc.uri_filter}`}
                                            active={tab_desc.uri_filter === roleFilter}
                                        >
                                            {tab_desc.tab_label}
                                        </Nav.Link>
                                        {/* </CSSTransition> */}
                                    </Nav.Item>
                                );
                            })}
                            <SearchBox />
                        </Nav>

                        <Tab.Content>
                            {tabs_desc.map((tab_desc) => {
                                return (
                                    <Tab.Pane
                                        key={'pane_' + tab_desc.tab_label}
                                        active={tab_desc.uri_filter === roleFilter}
                                    >
                                        <Table borderless>
                                            <thead>
                                                <tr>
                                                    {tab_desc.table_columns.map((col) => {
                                                        return (
                                                            <th key={tab_desc.tab_label + col.key}>
                                                                {col.renderHeader()}
                                                            </th>
                                                        );
                                                    })}
                                                </tr>
                                            </thead>
                                            <tbody>
                                                {users.map((user) => (
                                                    <tr key={tab_desc.tab_label + user.user_id}>
                                                        {tab_desc.table_columns.map((col) => {
                                                            return (
                                                                <td key={tab_desc.tab_label + user.user_id + col.key}>
                                                                    {col.renderCell(user)}
                                                                </td>
                                                            );
                                                        })}
                                                    </tr>
                                                ))}
                                            </tbody>
                                        </Table>
                                    </Tab.Pane>
                                );
                            })}
                        </Tab.Content>
                    </Tab.Container>
                </div>
            )}
        </ErrorBoundary>
    );
};

export default UserManagement;
