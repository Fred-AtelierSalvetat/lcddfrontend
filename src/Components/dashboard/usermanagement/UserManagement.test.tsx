import React from 'react';
import { expect } from 'chai';
import { mount } from 'enzyme';

import UserManagement from './UserManagement';

describe('<UserManagement /> ', () => {
    it('renders without crashing', () => {
        //const wrapper = mount(<UserManagement />);
    });
    it('proposes 3 tabs : "Admins", "Intervants" and "Utilisateurs"', () => {});
    it("'s 'Admins' tab selection should display the admin table with the specified columns", () => {});
    it("'s 'Admins' tab's table should allow 'Revenir intervenant' action", () => {});
    it("'s 'Admins' tab's table should allow 'Supprimer' action", () => {});
    it("'s 'Intervenants' tab selection should display the speakers table with the specified columns", () => {});
    it("'s 'Intervenants' tab's table should allow 'Promouvoir admin' action on validated speakers", () => {});
    it("'s 'Intervenants' tab's table should allow 'Valider candidat' action on speakers awaiting validation", () => {});
    it("'s 'Intervenants' tab's table should allow 'Valider candidat' action on speakers awaiting validation", () => {});
    it("'s 'Intervenants' tab's table should allow 'Supprimer' action", () => {});
    it("'s 'Utilisateurs' tab selection should display the users table with the specified columns", () => {});
    it("'s 'Utilisateurs' tab's table should allow 'Inviter Ã  intervenir' action on legal professional users", () => {});
    it("'s 'Utilisateurs' tab's table should allow 'Supprimer' action", () => {});

    it('should filter users according to role', () => {});
    it('should filter users according to search', () => {});
});
