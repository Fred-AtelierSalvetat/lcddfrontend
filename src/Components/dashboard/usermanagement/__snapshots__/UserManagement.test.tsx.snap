// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserManagement />  should match its reference snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <MemoryRouter
    initialEntries={
      Array [
        Object {
          "key": "invariableKeyForSnapshot",
          "pathname": "/users?tab=admin",
        },
      ]
    }
  >
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "canGo": [Function],
          "createHref": [Function],
          "entries": Array [
            Object {
              "hash": "",
              "key": "invariableKeyForSnapshot",
              "pathname": "/users?tab=admin",
              "search": "",
            },
          ],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "index": 0,
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "key": "invariableKeyForSnapshot",
            "pathname": "/users?tab=admin",
            "search": "",
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Route
        path={
          Array [
            "/users",
          ]
        }
      />
    </Router>
  </MemoryRouter>
</Provider>
`;
